### ISSUE TITLE
[Feature Request] Expose PromptValuesSupported to make extension of prompts easier

### ISSUE THREAD

**Issue by JustusGreiberORGADATA** [DuendeSoftware/Support#1142]
**Which version of Duende IdentityServer are you using?**
6.3.5

**Which version of .NET are you using?**
.NET 7

**Describe the ~bug~feature**

Right now we are looking at implementing a custom prompt type for our OpenID Connect flow. In our current IdentityServer version this seems possible by implementing `ICustomAuthorizeRequestValidator` and reinstating prompt values. The experience is not optimal, because the server prints conflicting log messages (first "prompt value not supported" later "prompt value reinstated"), but it works.

We now noticed that with an upgrade to IdentityServer 7 this probably won't work. This is because unknown prompt values were changed to cause errors in https://github.com/DuendeSoftware/IdentityServer/pull/1331 and errors cause an early return in the `AuthorizeRequestValidator`: https://github.com/DuendeSoftware/IdentityServer/blob/7.0.1/src/IdentityServer/Validation/Default/AuthorizeRequestValidator.cs#L121-L133

We also noticed, that the supported prompts are already defined in `UserInteractionOptions` https://github.com/DuendeSoftware/IdentityServer/blob/7.0.1/src/IdentityServer/Configuration/DependencyInjection/Options/UserInteractionOptions.cs#L134-L138. Changing these would allow a much simpler implementation of custom prompts. Additionally it would cause the prompt to show up in the discovery document without any further changes.
We can't set them though, because they are marked as internal.

Would it be possible to expose the PromptValuesSupported property?

**Comment by JustusGreiberORGADATA** [comment]
Hi,

> The spec does not include wording that this set of defined values is allowed to be extended

Interesting. I did interpret the section "If an OP receives a prompt value outside the set defined above _that it does not understand_" like "an implementation could understand other values outside the defined set". But TBF that is certainly not as clear/unambiguous as for example 5.1 Additional Claims (https://openid.net/specs/openid-connect-core-1_0.html#AdditionalClaims).

In the past, we also used the same `ICustomAuthorizeRequestValidator` to implement `prompt=create` while it was still a draft and not supported by IdentityServer. 

> What new prompt value are you looking to add?

In our application, the user can log in with the context of his organization/employer (Think "I, Justus Greiber, log in as an employee of ORGADATA AG." vs. "I, Justus Greiber, login as myself."). The prompt we are looking to add is one that shows this selection again. We could maybe use `select_account` for this, but we want to avoid this so that it can be used for "actual" multi-session support in the future.

**Comment by brockallen** [comment]
Hmm... ok. Well, for other customers I think they have used the "tenant" acr_values that we support. Would this help/work for you?

https://docs.duendesoftware.com/identityserver/v7/reference/endpoints/authorize/#optional-parameters
https://docs.duendesoftware.com/identityserver/v7/reference/services/interaction_service/#authorizationrequest

**Comment by JustusGreiberORGADATA** [comment]
As far as I understood the acr_values tenant mechanism in the past, it is more about the client telling the OpenID provider, what kind of tenant the client or maybe the user belongs to (`acr_values: ["tenant:ORGADATA"]`).

Our use case is that the user actively decides about the context and that the context is specific to the user. If I log in with the context of my employment at ORGADATA, it is specifically that employee-employer relation (ORGADATA:JustusGreiber). Data might be queried for exactly that relation, e.g., a license assigned to this relation by my employer.

If I have never used the client before, the client does not know about either me or my employer and can't send any tenant name/id. If I used the client with "ORGADATA:JustusGreiber" in the past, the client might know about this relation, but it does not know about my other employment at, e.g., DuendeSoftware (DuendeSoftware:JustusGreiber). This is exactly where the prompt comes into play. The client needs to tell the OpenId provider, "the user wants to login with a different employee relation I do not know about yet".

I guess it is not impossible to implement something like this with `acr_values`, one could use a special tenant to signal this, e.g., `acr_values: ["tenant:force_selection"]`. It is not our preferred solution because the mechanism thematically fits better with the other prompt values. For the same reason, we decided against a custom parameter like `orgadata_enhanced_prompts`, which would also have been a solution.

**Comment by brockallen** [comment]
I see -- so your user is the one deciding? How does the client know to pass the right value (regardless of the delivery mechanism)? Does it prompt the user prior to triggering OIDC to log them in? Does it always need to do this (for an anonymous user)? Does it ever change during the user's session at the client?

**Comment by brockallen** [comment]
Also of interest... the original 2014 (pre-errata) version has slightly different wording on this parameter: https://openid.net/specs/openid-connect-core-1_0-final.html#AuthRequest.

**Comment by JustusGreiberORGADATA** [comment]
> I see -- so your user is the one deciding? 

Yes

> How does the client know to pass the right value (regardless of the delivery mechanism)? Does it prompt the user prior to triggering OIDC to log them in?
> Does it ever change during the user's session at the client?


The main use case for this prompt is that the user is already logged in to the client with, for example, "ORGADATA:JustusGreiber".

We then show a drop-down-style UI element in the client to switch context. The client can show other employee-employeer relations it already knows about in this dropdown. If a known relation is chosen, we send a context hint during the authorization flow (similar to login_hint). No prompt is sent if the user chooses a context the client already knows.

If I am logged in with "ORGADATA:JustusGreiber" and used the client with "DuendeSoftware:JustusGreiber" and "Myself:JustusGreiber" in the past, it might look something like this:

Current Context: ORGADATA:JustusGreiber 


````
|=============================|
|SELECT NEW CONTEXT        🔽|
|                             |
|=============================|
|Myself:JustusGreiber         |
|                             |
|DuendeSoftware:JustusGreiber |
|                             |
|-----------------------------|
| + Connect new context +     |
|                             |
|-----------------------------|
````

The last button is what triggers the prompt. The switch will happen inside the same IdentityServer session, but it requires the user to create a new consent specific to that context and client. Additionally, users might use different contexts in different client applications at the same time (using the same IdentityServer session).

> Does it always need to do this (for an anonymous user)?

In the first authorization flow for a client for the current IdentityServer session the context selection will always appear, because you need an initial context to get started. The prompt is only for re-prompting the user during his active session at the client (much like the `login`, `select_account` and `consent` prompts). 
We are not preventing the client from sending this prompt for an anonymous user. But it won't change anything.

**Comment by JustusGreiberORGADATA** [comment]
The "connect new context flow" is a user interaction that is only available in the IdentityProvider itself. Because of the requirement for consent, the client is not allowed to access any data about the users other contexts beforehand, so it can't facilitate this interaction itself.

**Comment by brockallen** [comment]
@josephdecock Let's discuss

**Comment by josephdecock** [comment]
@JustusGreiberORGADATA - I've created the issue linked above in the main repo. We haven't yet made our plans for what will be included in the 7.1 release, but this looks like a good candidate for work that would typically get included in a minor release. For now, I'll close this issue in favor of that one.

**Comment by josephdecock** [comment]
@JustusGreiberORGADATA - just wanted to update you that today's release of identityserver 7.0.7 includes support for customizing the supported prompt modes. Please let us know how it works for you!
